### ISSUE TITLE
Create WebExtension Experiments and a demo VimFx extension

### ISSUE THREAD

**Issue by lydell** [lydell/webextension-keyboard#1]
- [ ] `browser.keyboard.onKey`, `browser.keyboard.onKeyPreventable` and `browser.keyboard.KeyboardEvent`
  - [x] Template
  - [ ] Demo implementation
- [x] `browser.keyboardShortcutActions` (although I had to name the demo API `browser.keyboardshortcutactions`, because I never got API names with uppercase letters in them to work.)
- [ ] "Escaping" from UI elements

**Comment by lydell** [comment]
For `browser.keyboard`, I think we need to:

- Specify the KeyboardEvent sent to the listeners as `browser.keyboard.KeyboardEvent`.
- Use the following manifest:

```json
[
  {
    "namespace": "keyboard",
    "permissions": ["experiments.keyboard"],
    "types": [
      {
        "id": "KeyboardEvent",
        "type": "object",
        "properties": {
          "type": {
            "type": "string",
            "description": "The string \"keydown\"."
          },
          "key": {
            "type": "string",
            "description": "The value of the keys or keys pressed by the user."
          },
          "etc": {
            "type": "todo",
            "description": "todo"
          }
        }
      }
    ],
    "events": [
      {
        "name": "onKey",
        "type": "function",
        "description": "Fired when the user presses a key, and the WebExtension is allowed to handle that key press.",
        "parameters": [
          {
            "name": "name",
            "$ref": "KeyboardEvent",
            "description": "The event object for the key that was pressed."
          }
        ]
      },
      {
        "name": "onKey",
        "type": "function",
        "description": "Fired when the user presses a key, and the WebExtension is allowed to handle that key press, and not if the web page has called `event.preventDefault()`.",
        "parameters": [
          {
            "name": "name",
            "$ref": "KeyboardEvent",
            "description": "The event object for the key that was pressed."
          }
        ]
      }
    ]
  }
]
```

**Comment by lydell** [comment]
Pushed some WIP progress today.

**Comment by cmcaine** [comment]
I'm experimenting with re-implementing vimperator with webextensions and would like to use your WIP webextension-keyboard, but I can't work out how.

Could you help me out?

**Comment by lydell** [comment]
It's so much a work in progress that it doesn't even work yet :)

**Comment by lydell** [comment]
Note that this won't receive any more progress since I've decided to abandon the project. See the readme for more information.

**Comment by cmcaine** [comment]
Thanks for updating me. I would have just happily waited for weeks :)

The Firefox devs seem willing to have an API like this and I think it's still useful for my use case.

Is there anything beyond the published stuff in this repo that you could pass on if I were to look into developing this myself?

**Comment by lydell** [comment]
I don't think so. If you do decide to develop something like this, I wish you the best of luck! Also remember to mention your work here, because I and people who watch this repo would be very interested :)

**Comment by cmcaine** [comment]
Thanks, and I will do if I get anywhere.

**Comment by cmcaine** [comment]
An update: I (finally) managed to get in touch with some Firefox developers about this two weeks ago and have made a start on a prototype API this week. I expect to have the first prototype done on my next evening free, maybe by the end of the weekend.

I tried to implement the API described here first, but the return `true` from the listener to suppress keyevents is problematic: The API has to live in the Firefox main process while webextensions live in their own processes and there's no real support for synchronous calls between processes.

Code to clarify:

```js
// Passed to window.addEventListener, calls each keydownListener.
this.dispatcher = (ke) => {
  for (let listener of this.keydownListeners) {
    // There's no easy way to make listener return anything but a Promise, so
    // this doesn't work.
    requests_suppression = listener(ke.key);
    // We could suppress all events, then re-dispatch the ones that listeners
    // don't object to, but that's likely to run into some bugs and is pretty
    // messy.
    if (requests_suppression) {
      ke.preventDefault();
    }
  }
```

So what I'm proposing instead is that callers must specify rules in advance that dictate what keyevents will be suppressed and that these rules should be in some declarative form and could depend on:

 - Attributes of the current keyevent
 - Previous keyevents
 - Simple logic and arithmetic
 - *Maybe* the rules are allowed to save and restore small amounts of state.
 - That's it.

The rules can be changed by extensions, but if keys are pressed quickly after each other, then it's possible that a rule change triggered by the first key press will not be enforced by the API until some arbitrary number of keypresses have been handled, so extensions should try to avoid dynamically changing the suppression rules.

I recognise that this is problematic, but I think it's something we can work with. I'm fairly sure Vimperator/VimFx type functionality is easily doable with this system thanks to them modes we like so much. Something more emacsy (see issue #3) is also possible, but the rules will get a bit more complicated.

This is a RFC :)

I don't use emacs much, so I'd really appreciate if anyone more familiar with it would be willing to try and come up with some rules that fit my restrictions above that would provide the right kind of suppression behaviour.

I'd also be interested in hearing about any small, restricted languages that I could have the rules expressed in. I would prefer the rules language/system to only allow short terminating programs to be expressed.

**Comment by ncalexan** [comment]
> I don't use emacs much, so I'd really appreciate if anyone more familiar with it would be willing to try and come up with some rules that fit my restrictions above that would provide the right kind of suppression behaviour.

I don't have much bandwidth to help here, but I will add a few notes.

> So what I'm proposing instead is that callers must specify rules in advance that dictate what keyevents will be suppressed and that these rules should be in some declarative form and could depend on:
> 
>     Attributes of the current keyevent
>     Previous keyevents
>     Simple logic and arithmetic
>     Maybe the rules are allowed to save and restore small amounts of state.
>     That's it.
> 
> The rules can be changed by extensions, but if keys are pressed quickly after each other, then it's possible that a rule change triggered by the first key press will not be enforced by the API until some arbitrary number of keypresses have been handled, so extensions should try to avoid dynamically changing the suppression rules.

I think taking a page from Emacs might simplify this considerably.  But -- I'm not sure that this simplification will handle vi _modes_ well.

In Emacs, keys are mapped via keymaps.  Keymaps associate keys to either commands (basically what Web Extensions provide now) or to further keymaps.  See https://www.gnu.org/software/emacs/manual/html_node/elisp/Keymaps.html for details, but in essence you can say:
```js
fn cut() { ... };
fn undo() { ... };

global_key_map = {
  'Ctrl-X': cut,
  'Ctrl-U': undo_key_map,
};
undo_key_map = {
  'Enter': undo,
};
```
That's enough to get multi-stroke key shortcuts, which is most of what I want.  In your list, that is the first two points (current key, previous key).

For vi modes, you might need to be able to programatically save and restore the global key map.  But I'd rather have a strictly data-driven structure like Emacs keymaps and a save/restore function than boolean logic or a custom DSL about repressing events.

Hope that helps!

**Comment by dumblob** [comment]
> But I'd rather have a strictly data-driven structure like Emacs keymaps and a save/restore function than boolean logic or a custom DSL about repressing events.

I second this recommendation.

**Comment by cmcaine** [comment]
Thanks very much for your comments.

I was thinking of doing a finite state machine that transitions based on keyevent, with each state having a white or blacklist, which seems essentially the same, so glad to see rough agreement :)

As for vi-mode: modes are triggered by pressing keys, so the same structure works.

**Comment by cmcaine** [comment]
Here's a first working version of my API: https://github.com/cmcaine/keyboard-api

The more complicated event suppression logic stuff isn't developed yet as I'm still ironing out other issues.

**Comment by eejdoowad** [comment]
Hey Colin, thanks for your effort. One concern I have is that the declarative API you've devised is a little bit complicated.

Since you seem familiar with the firefox internals at this stage, can you critique this idea:

Extensions can add/remove keydown and keyup listeners in their background page. Events are passed to these listeners before ANYTHING else (including firefox's chrome). The return value of these listeners determines whether the event
1) goes down the usual path to Firefox's Chrome and then to the page
2) goes directly to the page
3) goes no where at all

One obvious danger in this design is that an extension can break all keyboard shortcuts. Another is that an extra pass to the background page process will reduce responsiveness.

**Comment by cmcaine** [comment]
That was more or less the original API proposal, and, if Firefox people permit it, it's what I'll do, but last time I spoke to one of them about it, they weren't happy about the idea and this more complicated system was the only way it could work. It's possible that I will be able to persuade whoever takes over mentorship of the bug to support the simpler system, but I certainly wouldn't count on it.

On the upside, I'm developing the kind of key sequence parser that the declarative API would require over in [tridactyl](https://github.com/cmcaine/tridactyl), so even though there's been no movement on the API repo, work is being done.

I think you're probably clued up enough to know the gist of what's required, but for public interest:

At present, there are three relevant scripts running for key events: chrome, frame, and webextension. These scripts can only communicate by passing messages. Firefox's current policy is that non-chrome scripts may not block the chrome script (possibly they make some exceptions for this).

To do as you suggest, the chrome script has to put each keyevent into a queue, then generally asynchronously talk to the frame script, then the webextension script, then decide what to do with the event. And, presumably there needs to be some logic in there to ensure that the ordering of events is maintained, even if the ordering of responses from either the frame or webextension are off.
